File: src/test/java/ar/edu/unq/spring/application/service/InventarioServiceTest.java
package ar.edu.unq.spring.application.service;

import ar.edu.unq.spring.application.port.input.InventarioUseCase;
import ar.edu.unq.spring.application.port.input.PersonajeUseCase;
import ar.edu.unq.spring.domain.model.Item;
import ar.edu.unq.spring.domain.model.exception.MuchoPesoException;
import ar.edu.unq.spring.domain.model.exception.NombreDePersonajeRepetidoException;
import ar.edu.unq.spring.domain.model.Personaje;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;

@SpringBootTest
public class InventarioServiceTest {

    @Autowired
    private InventarioUseCase inventarioUseCase;

    @Autowired
    private PersonajeUseCase personajeUseCase;

    private Personaje maguin;
    private Personaje debilucho;
    private Item baculo;
    private Item tunica;

    @BeforeEach
    public void prepararTest() {
        tunica = inventarioUseCase.crearItem(new Item("Tunica", 100));
        baculo = inventarioUseCase.crearItem(new Item("Baculo", 50));
        
        maguin = personajeUseCase.crearPersonaje(new Personaje("Maguin", 10, 70));
        debilucho = personajeUseCase.crearPersonaje(new Personaje("Debilucho", 1, 1000));
    }

    @Test
    public void testRecogerItem() {
        inventarioUseCase.agregarItemAPersonaje(maguin.getId(), baculo);

        Personaje maguinActualizado = personajeUseCase.obtenerPersonajePorId(maguin.getId());
        Assertions.assertEquals("Maguin", maguinActualizado.getNombre());
        Assertions.assertEquals(1, maguinActualizado.getInventario().size());

        Item baculoRecuperado = maguinActualizado.getInventario().iterator().next();
        Assertions.assertEquals("Baculo", baculoRecuperado.getNombre());
        Assertions.assertSame(baculoRecuperado.getDueño(), maguinActualizado);
    }

    @Test
    public void testObtenerTodosLosItems() {
        // Primero agregamos los items a algún personaje para que estén en el sistema
        inventarioUseCase.agregarItemAPersonaje(maguin.getId(), baculo);
        inventarioUseCase.agregarItemAPersonaje(debilucho.getId(), tunica);
        
        // Verificar obtención de inventario
        List<Item> itemsMaguin = inventarioUseCase.obtenerInventarioDePersonaje(maguin.getId());
        List<Item> itemsDebilucho = inventarioUseCase.obtenerInventarioDePersonaje(debilucho.getId());
        
        Assertions.assertEquals(1, itemsMaguin.size());
        Assertions.assertEquals(1, itemsDebilucho.size());
    }

    @Test
    public void testExcepcionPorMuchoPeso() {
        // Añadir el báculo a maguin
        inventarioUseCase.agregarItemAPersonaje(maguin.getId(), baculo);
        
        // Intentar añadir la túnica (que pesa 100) debería fallar porque excede el límite
        MuchoPesoException exception = Assertions.assertThrows(
            MuchoPesoException.class, 
            () -> inventarioUseCase.agregarItemAPersonaje(maguin.getId(), tunica)
        );
        
        String mensajeError = exception.getMessage();
        Assertions.assertTrue(mensajeError.contains("Maguin"));
        Assertions.assertTrue(mensajeError.contains("Tunica"));
    }

    @Test
    public void testNombreDePersonajeTieneQueSerUnico() {
        // Intentar crear otro personaje con el mismo nombre
        Personaje otroMaguin = new Personaje("Maguin", 10, 70);

        NombreDePersonajeRepetidoException exception = Assertions.assertThrows(
            NombreDePersonajeRepetidoException.class, 
            () -> personajeUseCase.crearPersonaje(otroMaguin)
        );
        
        Assertions.assertTrue(exception.getMessage().contains("Maguin"));
    }

    @AfterEach
    public void limpiarDatos() {
        // Esto no deberia estar en el useCase, no? Lo dejo aca por comodidad
        personajeUseCase.eliminarTodosLosPersonajes();
    }
} 

File: src/test/kotlin/ar/edu/unq/spring/InventarioServiceTest.kt
package ar.edu.unq.spring

import ar.edu.unq.spring.modelo.Item
import ar.edu.unq.spring.modelo.Personaje
import ar.edu.unq.spring.service.InventarioService
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension

@ContextConfiguration(classes = [InventarioServiceTestConfig::class])
@ExtendWith(SpringExtension::class)
@DataJpaTest
@TestInstance(PER_CLASS)
class InventarioServiceTest {

    lateinit var maguin: Personaje
    lateinit var debilucho: Personaje
    lateinit var baculo: Item
    lateinit var tunica: Item

    @Autowired
    lateinit var service: InventarioService

    @BeforeAll
    fun prepare() {

        tunica = Item("Tunica", 100)
        baculo = Item("Baculo", 50)

        service.guardarItem(tunica)
        service.guardarItem(baculo)

        maguin = Personaje("Maguin")
        maguin.pesoMaximo = 70
        maguin.vida = 10
        service.guardarPersonaje(maguin)


        debilucho = Personaje("Debilucho")
        debilucho.pesoMaximo = 1000
        debilucho.vida = 1
        service.guardarPersonaje(debilucho)
    }

    @Test
    fun testRecoger() {
        service.recoger(maguin.id, baculo.id)

        val maguito = service.recuperarPersonaje(maguin.id)
        Assertions.assertEquals("Maguin", maguito?.nombre)

        Assertions.assertEquals(1, maguito?.inventario?.size)

        val baculo = maguito?.inventario?.iterator()?.next()
        Assertions.assertEquals("Baculo", baculo?.nombre)

        Assertions.assertSame(baculo?.owner, maguito)
    }

    @Test
    fun testGetAll() {
        val items = service.allItems()

        Assertions.assertEquals(2, items.size.toLong())
        Assertions.assertTrue(items.contains(baculo))
    }

    @Test
    fun testGetMasPesados() {
        val items = service.getMasPesados(10)
        Assertions.assertEquals(2, items.size.toLong())

        val items2 = service.getMasPesados(80)
        Assertions.assertEquals(1, items2.size.toLong())
    }

    @Test
    fun testGetItemsDebiles() {
        var items = service.getItemsPersonajesDebiles(5)
        Assertions.assertEquals(0, items.size.toLong())

        service.recoger(maguin.id, baculo.id)
        service.recoger(debilucho.id, tunica.id)

        items = service.getItemsPersonajesDebiles(5)
        Assertions.assertEquals(1, items.size.toLong())
        Assertions.assertEquals("Tunica", items.iterator().next().nombre)

    }

    @Test
    fun testGetMasPesado() {
        val item = service.heaviestItem()
        Assertions.assertEquals("Tunica", item.nombre)
    }
}